
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ChIP-seq workflow &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Integrative workflows" href="integrative.html" />
    <link rel="prev" title="Detailed documentation of RNA-Seq downstream" href="rnaseq-rmd.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="workflows.html">Overview of workflows</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="workflows.html#primary-analysis-workflows">Primary analysis workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="workflows.html#integrative-analysis-workflows">Integrative analysis workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="workflows.html#features-common-to-workflows">Features common to workflows</a></li>
<li class="toctree-l2"><a class="reference internal" href="workflows.html#next-steps">Next Steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration details</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
  <li><a href="workflows.html">Overview of workflows</a><ul>
      <li>Previous: <a href="rnaseq-rmd.html" title="previous chapter">Detailed documentation of RNA-Seq downstream</a></li>
      <li>Next: <a href="integrative.html" title="next chapter">Integrative workflows</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="chip-seq-workflow">
<span id="chipseq"></span><h1>ChIP-seq workflow<a class="headerlink" href="#chip-seq-workflow" title="Permalink to this headline">Â¶</a></h1>
<p>The ChIP-seq workflow starts with raw FASTQ files, does various QC steps,
aligns and prepares BAM and bigWig files, performs peak-calling, and combines
everything together into a track hub for visualization.</p>
<p>Specifically, the workflow does the following:</p>
<blockquote>
<div><ul class="simple">
<li><p>trim reads with cutadapt</p></li>
<li><p>map reads with Bowtie2</p></li>
<li><p>run FastQC on raw, trimmed, and aligned reads</p></li>
<li><p>Remove multimappers (samtools) and duplicates (Picard MarkDuplicates)</p></li>
<li><p>fastq_screen on multiple configured genomes to look for evidence of
cross-contamination</p></li>
<li><p>QC aggregation using MultiQC, along with a custom table for library sizes</p></li>
<li><p>merge technical replicates and then re-deduplicate</p></li>
<li><p>create bigWigs from unique, no-dups BAM files</p></li>
<li><p>optionally merge bigWigs to create one signal track for all replicates</p></li>
<li><p>run deepTools plotFingerprint on grouped IP and input for QC and
evaluation of enrichment</p></li>
<li><p>call peaks using macs2, spp, and/or sicer, with support for multiple
peak-calling runs using different parameters to assist with assessing
performance and to help make decisions for downstream analysis</p></li>
<li><p>optionally run a template diffBind RMarkdown file used for differential binding analysis</p></li>
<li><p>convert BED files into bigBed (or bigNarrowPeak where possible)</p></li>
<li><p>build and optionally upload a track hub of bigWigs and bigBeds to
visualize peak-calling in UCSC Genome Browser</p></li>
</ul>
</div></blockquote>
<p>To configure a ChIP-seq experiment, see <a class="reference internal" href="config-yaml.html#config-yaml"><span class="std std-ref">Config YAML</span></a>.</p>
<img alt="_images/chipseq.png" src="_images/chipseq.png" />
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/chipseq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>