
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>“Figures” workflow &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Colocalization workflow" href="colocalization.html" />
    <link rel="prev" title="“External” workflow" href="external.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Initial setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Running the tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="deploy.html">Deploying <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> and staying up-to-date</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">Downstream RNA-seq in R</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="external.html">“External” workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">“Figures” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocalization.html">Colocalization workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster.html">Running on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="autodoc.html">Module documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="external.html" title="previous chapter">“External” workflow</a></li>
      <li>Next: <a href="colocalization.html" title="next chapter">Colocalization workflow</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="figures-workflow">
<span id="figures"></span><h1>“Figures” workflow<a class="headerlink" href="#figures-workflow" title="Permalink to this headline">¶</a></h1>
<p>This workflow is a working example of how you would tie together output from
RNA-seq, ChIP-seq, and “external” workflows to automate figure creation and
formally link together the dependencies of each figure all the way back to the
original fastq files. If any changes are made upstream, they will trigger
downstream rules to re-run as needed.</p>
<p>For example, if a new GTF annotation file comes out, you would change the URL
in the config, and then re-run the figures workflow. All RNA-seq jobs that
depended on some way on the GTF file will be re-run. This will include the
feature counts and downstream RNA-seq analysis, but will <em>not</em> re-run any of
the jobs like trimming or mapping that do not depend on the GTF. In addition,
any figures that depended in some way on that GTF file will also be re-run.</p>
<p>To provide a sufficiently complex example that can be used in real-world
applications, this workflow currently:</p>
<blockquote>
<div><ul class="simple">
<li><p>counts the number of peaks in all configured peak-calling runs and stores
the output in a TSV report (work is performed in the script
<code class="docutils literal notranslate"><span class="pre">scripts/peak_count.py</span></code>)</p></li>
<li><p>identifies peaks at promoters of genes, reports a summary of how many
peaks from each run were found in a promoter, and creates BED files of
these subsets (<code class="docutils literal notranslate"><span class="pre">scripts/peaks_at_promoters.py</span></code>)</p></li>
<li><p>builds the DAG image for the ChIP-seq and RNA-seq workflows</p></li>
<li><p>symlinks over the ChIP-seq peaks and RNA-seq differential expression results</p></li>
<li><p>extracts the text from the README.txt files created by the scripts
(usually from their docstrings), and compiles them into a summary report</p></li>
<li><p>the <code class="docutils literal notranslate"><span class="pre">figures</span></code> directory can then be zipped up and distributed to collaborators</p></li>
</ul>
</div></blockquote>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/figures.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>