
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Changelog &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Initial setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Running the tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">Downstream RNA-seq in R</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="external.html">“External” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="figures.html">“Figures” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocalization.html">Colocalization workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster.html">Running on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="autodoc.html">Module documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="changelog">
<h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h1>
<div class="section" id="v1-0">
<h2>v1.0<a class="headerlink" href="#v1-0" title="Permalink to this headline">¶</a></h2>
<p>First full release.</p>
</div>
<div class="section" id="v1-0-1">
<h2>v1.0.1<a class="headerlink" href="#v1-0-1" title="Permalink to this headline">¶</a></h2>
<p>Bugfixes, last release before references changes.</p>
<div class="section" id="infrastructure">
<h3>Infrastructure<a class="headerlink" href="#infrastructure" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Transition to CircleCI for testing</li>
<li>Use production settings by default; see <span class="xref std std-ref">note-on-test-settings</span> for
more.</li>
<li>lots o’ docs</li>
<li>new <code class="docutils literal notranslate"><span class="pre">include/references_configs</span></code> to help organize references. These are
currently not used by the workflows directly.</li>
<li>bugfix: use additional options when uncompressing downloaded reference files
(<code class="docutils literal notranslate"><span class="pre">--no-same-owner</span></code> for <code class="docutils literal notranslate"><span class="pre">tar</span></code>, <code class="docutils literal notranslate"><span class="pre">-f</span></code> for <code class="docutils literal notranslate"><span class="pre">gunzip</span></code>)</li>
<li>additional dependencies in the top-level environment to support the
additional features in rnaseq.Rmd and track hubs.</li>
<li>colocalization workflow, external workflow, figures workflow to demonstrate
vertical integration</li>
</ul>
</div>
<div class="section" id="rna-seq">
<h3>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>remove kallisto indexing, use salmon</li>
<li>improvements to how chipseq sampletables are parsed (with more informative
error messages)</li>
<li>run preseq for RNA-seq library complexity QC</li>
<li>support for merging bigwigs</li>
<li>rule for symlinking stranded bigwigs</li>
<li>featureCounts is now run in all three strandedness modes, and results
incorporated into MultiQC as separate modules.</li>
<li>new <code class="docutils literal notranslate"><span class="pre">downstream/helpers.Rmd</span></code> which factors out a lot of the work previously
done in <code class="docutils literal notranslate"><span class="pre">rnaseq.Rmd</span></code> into separate functions.</li>
<li>track hub building respects new sense/antisense bigwig symlinks</li>
</ul>
</div>
<div class="section" id="downstream-rnaseq-rmd">
<h3><code class="docutils literal notranslate"><span class="pre">downstream/rnaseq.Rmd</span></code><a class="headerlink" href="#downstream-rnaseq-rmd" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>AnnotationHub uses cache dir that will not clobber default home directory cache</li>
<li>use varianceStabilizingTransform instead of rlog</li>
<li>print a size factors table</li>
<li>use multiple cores for computationally expensive DESeq2 operations</li>
<li>using separate lists for results, dds objects, and nice labels for automated
plots for each contrast</li>
<li>UpSet plots for comparing gene lists across contrasts</li>
<li>DEGpattern plots for showing clusters of expression patterns (from the
DEGreport package)</li>
<li>attach normalized counts per sample and per factor (parsed from the model
used for the contrast) as well as TPM estimates to the results tables</li>
<li>trim the labels in GO enrichment plots when too long</li>
</ul>
</div>
<div class="section" id="chip-seq">
<h3>ChIP-seq<a class="headerlink" href="#chip-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>sicer for chipseq domain calling</li>
<li>pin snakemake &lt;4.5.0 so that subworkflows behave correctly</li>
<li>chipseq peak-calling rules (and therefore wrappers) now expect a chromsizes
file as input</li>
<li>bigbed files for narrowPeak and broadPeak files are created correctly
depending on their format</li>
<li>run multiBigWigSummary and plotCorrelation from deepTools for ChIP-seq QC</li>
<li>ChIP-seq track hub generation script</li>
</ul>
</div>
<div class="section" id="both-rna-seq-and-chip-seq">
<h3>Both RNA-seq and ChIP-seq<a class="headerlink" href="#both-rna-seq-and-chip-seq" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>update deeptools calls to reflect &gt;v3.0 syntax</li>
<li>support for SRA run tables so it’s trivial to re-run experiments
in SRA</li>
<li>multiple FastQC runs are shown separately in MultiQC output</li>
</ul>
</div>
</div>
<div class="section" id="v1-1">
<h2>v1.1<a class="headerlink" href="#v1-1" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p class="first"><code class="docutils literal notranslate"><span class="pre">include/reference_configs</span></code> contains example config files for various
species that can be pasted into individual workflows’ <code class="docutils literal notranslate"><span class="pre">references:</span></code>
section.</p>
</li>
<li><p class="first"><cite>fastq_screen</cite> is now configured via <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>. This reduces the need
to edit the Snakefile and coordinate between the config and the fastq_screen
rule. Now everything is done within the config file.</p>
</li>
<li><p class="first">In the config file, <code class="docutils literal notranslate"><span class="pre">assembly</span></code> is now <code class="docutils literal notranslate"><span class="pre">organism</span></code>. The change is backwards
compatible, but a DeprecationWarning is raised if <code class="docutils literal notranslate"><span class="pre">assembly:</span></code> is still
used, and it is then automatically copied over to <code class="docutils literal notranslate"><span class="pre">organism</span></code></p>
</li>
<li><p class="first">using CircleCI for continuous integration testing</p>
</li>
<li><p class="first">the default settings in Snakefiles are for real-world use, rather than for
testing. This reduces the amount of editing necessary before running actual
data. See <a class="reference internal" href="tests.html#test-settings"><span class="std std-ref">A note about test settings</span></a> for the extra step to take when testing
locally.</p>
</li>
<li><p class="first">All Snakefiles now have the string <code class="docutils literal notranslate"><span class="pre">#</span> <span class="pre">NOTE</span></code> to indicate settings that may
need to be changed depending on the experiment.</p>
</li>
<li><p class="first">Documentation overhaul to bring everything up to v1.1, including Sphinx
autodocs on the <code class="docutils literal notranslate"><span class="pre">lib</span></code> module.</p>
</li>
<li><p class="first">patterns no longer use <code class="docutils literal notranslate"><span class="pre">{sample_dir}</span></code>, <code class="docutils literal notranslate"><span class="pre">{agg_dir}</span></code>, etc placeholders that
need to be configured in the config YAML. Instead, these directories are
hard-coded directly into the patterns. This simplifies the config files,
simplifies the patterns, and removes one layer of disconnect between the
filenames and how they are determined.</p>
</li>
<li><p class="first">additional section in configs for averaging bigWigs in a flexible manner</p>
</li>
<li><p class="first">removed 4C workflow since it used 4c-ker</p>
</li>
<li><p class="first">ChIP-seq and RNA-seq can now run directly on sample tables from SRA’s run
browser (fastqs will be downloaded with fastq-dump, etc).</p>
</li>
<li><p class="first">pinning to deepTools &gt;3.0 to use new command line args</p>
</li>
<li><p class="first">ChIP-seq now supports SICER for peak-calling.</p>
</li>
<li><p class="first">ChIP-seq multibigwigsummary and heatmaps for QC and clustering of samples</p>
</li>
<li><p class="first">ChIP-seq track hub with peaks and signal</p>
</li>
<li><p class="first">Peak caller wrappers now handle their own conversion to bigBed</p>
</li>
<li><p class="first">macs2 and sicer can accept mappable genome size overrides</p>
</li>
<li><p class="first">Improvements to MultiQC config so that multiple module runs (e.g., FastQC on
raw, trimmed, aligned) show up separately in the stats table.</p>
</li>
<li><p class="first">Added colocalization workflow, external workflow, and figures workflow.</p>
</li>
<li><p class="first">RNA-seq now runs all three strandedness for featureCounts, and reports them
all in MultiQC so you can check protocol expectations.</p>
</li>
<li><p class="first">RNA-seq now runs preseq for QC</p>
</li>
<li><p class="first">RNA-seq now symlinks “pos” and “neg” bigWigs, which describe how reads map to
the <em>reference</em> to “sense” and “antisense” bigWigs, which describe the
originating RNA. This makes it easy to swap strands depending on protocol.</p>
</li>
<li><p class="first">RNA-seq track hub with signal</p>
</li>
<li><p class="first">RNA-seq downstream:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">downstream/help_docs.Rmd</span></code> can be included for first-time users to
describe the sections of the RNA-seq analysis</li>
<li>factored out <code class="docutils literal notranslate"><span class="pre">downstream/helpers.Rmd</span></code> into a separate file that is
included into the main Rmd.</li>
<li>AnnotationHub is cached in <code class="docutils literal notranslate"><span class="pre">include</span></code> dir of repo to avoid clobbering
any locally-installed versions.</li>
<li>UpSet plot example to show how to compare results from multiple contrasts</li>
<li>Easy swapping of which strand to use from the three featureCounts runs
performed by the workflow</li>
<li>Be explicit about using DESeq2::lfcShrink as is now the default in recent
DESeq2 versions</li>
<li>DEGpatterns plots</li>
</ul>
</div></blockquote>
</li>
</ul>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
      |
      <a href="_sources/full-changelog.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>