
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Getting started &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Guide to file hierarchy" href="guide.html" />
    <link rel="prev" title="Introduction" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="toc.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-a-project">Setting up a project</a></li>
<li class="toctree-l2"><a class="reference internal" href="#deploying-lcdb-wf">1. Deploying lcdb-wf</a></li>
<li class="toctree-l2"><a class="reference internal" href="#configure">2. Configure</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run">3. Run</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="guide.html">Guide to file hierarchy</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Testing the installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration details</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">RNA-Seq downstream analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="integrative.html">Integrative workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">For Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="toc.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Introduction</a></li>
      <li>Next: <a href="guide.html" title="next chapter">Guide to file hierarchy</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="getting-started">
<span id="id1"></span><h1>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h1>
<p>The only starting requirement is an installation of conda with the <a class="reference external" href="https://bioconda.github.io">bioconda</a> channel set up.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><cite>lcdb-wf</cite> is tested and heavily used on Linux.</p>
<p>It is likely to work on macOS as long as all relevant conda packages are
available for macOS – though this is not tested.</p>
<p>It will <strong>not</strong> work on Windows due to a general lack of support of Windows
in bioinformatics tools.</p>
</div>
<p>We use conda to install a separate environment for each project. This allows
long-running projects to keep using old versions of software if needed while
allowing newer versions in more recent projects.</p>
<p>To see the full list of software installed into these environments,
see the <code class="docutils literal notranslate"><span class="pre">requirements-*.txt</span></code> files at <a class="reference external" href="https://github.com/lcdb/lcdb-wf/">https://github.com/lcdb/lcdb-wf/</a>.</p>
<p>For a more detailed discussion of conda, see <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">Using conda and conda envs</span></a>.</p>
<div class="section" id="setting-up-a-project">
<span id="setup-proj"></span><h2>Setting up a project<a class="headerlink" href="#setting-up-a-project" title="Permalink to this headline">¶</a></h2>
<p>The general steps to use lcdb-wf in a new project are:</p>
<ol class="arabic simple">
<li><p>Deploy: from a clone of the lcdb-wf repository, run <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code></p></li>
<li><p>Configure: set up samples table for experiments and edit configuration file</p></li>
<li><p>Run: activate environment and run the Snakemake file either locally or on a cluster</p></li>
</ol>
</div>
<div class="section" id="deploying-lcdb-wf">
<span id="deploy"></span><h2>1. Deploying lcdb-wf<a class="headerlink" href="#deploying-lcdb-wf" title="Permalink to this headline">¶</a></h2>
<p>lcdb-wf is designed to be customized for each project, so setting up a new
project to use lcdb-wf means copying over the files you need. There is
a <code class="docutils literal notranslate"><span class="pre">deploy.py</span></code> script that does this for you.</p>
<p><code class="docutils literal notranslate"><span class="pre">deploy.py</span></code> sets up a project and removes additional files (like these docs
and testing files) that are not necessary for an actual project. The reason to use this script is so you end up with a cleaner project directory.</p>
<p>This script also writes a file to the destination called
<code class="docutils literal notranslate"><span class="pre">.lcdb-wf-deployment.json</span></code>. It stores the timestamp and details about what
commit was used to deploy it. This tracks provenance of the code, so you can always
figure out what lcdb-wf commit your deployment originally started from.</p>
<p>To deploy a copy, you first need to get a copy. Here are three ways to do that:</p>
<p>Clone a repo using git:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone git@github.com:lcdb/lcdb-wf.git

<span class="c1"># optionally check out a particular branch</span>
<span class="c1"># git checkout BRANCHNAME</span>
</pre></div>
</div>
<p>Or update an existing repo:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">BRANCH</span><span class="o">=</span>master  <span class="c1"># optionally change branch</span>
git checkout <span class="nv">$BRANCH</span>
git pull origin <span class="nv">$BRANCH</span>
</pre></div>
</div>
<p>No git? Download a zip file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">BRANCH</span><span class="o">=</span>master  <span class="c1"># optionally change branch</span>
wget https://github.com/lcdb/lcdb-wf/archive/<span class="nv">$BRANCH</span>.zip
unzip <span class="nv">$BRANCH</span>.zip -d lcdb-wf
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you want to run the tests then don’t deploy just yet – see
<a class="reference internal" href="tests.html#running-the-tests"><span class="std std-ref">Testing the installation</span></a> for details, and then come back here to deploy for
an actual project.</p>
</div>
<p>Now that you have a copy of the code, you can deploy it to your project
directory. Run the deploy script in the newly cloned/updated repo. For help,
use <code class="docutils literal notranslate"><span class="pre">-h</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python deploy.py -h
</pre></div>
</div>
<p>For example, to deploy the RNA-seq workflow to the <code class="docutils literal notranslate"><span class="pre">myproj</span></code> directory and
build the conda environments:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python deploy.py --flavor rnaseq --dest myproj --build-envs
</pre></div>
</div>
<p>Copying over the files is fast; building the conda environments may take a few
minutes.</p>
<p>See <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">Using conda and conda envs</span></a> for more details on these.</p>
</div>
<div class="section" id="configure">
<h2>2. Configure<a class="headerlink" href="#configure" title="Permalink to this headline">¶</a></h2>
<p>This step takes the most effort and the first time you set up a project it
will take some time to understand the configuration system.</p>
<ul class="simple">
<li><p>see <a class="reference internal" href="faqs.html#multiple-experiments"><span class="std std-ref">How do I handle multiple experiments in the same project?</span></a> for advice on how to handle multiple experiments that are intended to be analyzed together</p></li>
<li><p>see <a class="reference internal" href="conda.html#conda-envs"><span class="std std-ref">Using conda and conda envs</span></a> for details on conda environments</p></li>
<li><p>see <a class="reference internal" href="sampletable.html#sampletable"><span class="std std-ref">Sample tables</span></a> for how to write a sampletable, which includes where to find raw data and contains the associated metadata</p></li>
<li><p>see <a class="reference internal" href="config-yaml.html#config-yaml"><span class="std std-ref">Config YAML</span></a> for configuring each workflow</p></li>
</ul>
</div>
<div class="section" id="run">
<h2>3. Run<a class="headerlink" href="#run" title="Permalink to this headline">¶</a></h2>
<p>Activate the main environment, go to the workflow you want to run, and run the
following:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake --dryrun
</pre></div>
</div>
<p>If all goes well, this should print a list of jobs to be run.</p>
<p>You can run locally, but this is NOT recommended. To run locally, choose the
number of CPUs you want to use (here, 8).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you haven’t made any changes to the Snakefiles, be aware that the
default configuration needs a lot of RAM. For example, the MarkDuplicates
runs set 20 GB RAM for Java, and that’s for each job. Adjust the Snakefiles
accordingly if you don’t have enough RAM available (search for “Xmx” to
find the Java args that set memory).</p>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># run locally (not recommended)</span>
snakemake --use-conda -j <span class="m">8</span>
</pre></div>
</div>
<p>The recommended way is to run on a cluster. On NIH’s Biowulf cluster, the way
to do this is to submit the wrapper script as a batch job:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch ../../include/WRAPPER_SLURM
</pre></div>
</div>
<p>and then monitor the various jobs that will be submitted on your behalf. See
<a class="reference internal" href="config.html#cluster"><span class="std std-ref">Running on a cluster</span></a> for more details on this.</p>
<p>Other clusters will need different configuration, but everything is standard
Snakemake. The Snakemake documentation on <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cluster.html">cluster execution</a> and
<a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cloud.html">cloud execution</a> can be
consulted for running on your particular system.</p>
<p>You can typically run simultaneous workflows when they are in different directories; see
<a class="reference internal" href="workflows.html#workflows"><span class="std std-ref">Overview of workflows</span></a> for details.</p>
<div class="section" id="next-steps">
<h3>Next steps<a class="headerlink" href="#next-steps" title="Permalink to this headline">¶</a></h3>
<p>Next, we give a brief overview of the file hierarchy of <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> in the
<a class="reference internal" href="guide.html#guide"><span class="std std-ref">Guide to file hierarchy</span></a> page.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.5.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/getting-started.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>