
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Module lib.chipseq &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Module lib.patterns_targets" href="lib.patterns_targets.html" />
    <link rel="prev" title="Module lib.common" href="lib.common.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Initial setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Running the tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">Downstream RNA-seq in R</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="external.html">“External” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="figures.html">“Figures” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocalization.html">Colocalization workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster.html">Running on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="autodoc.html">Module documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="lib.common.html">Module <code class="docutils literal notranslate"><span class="pre">lib.common</span></code></a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Module <code class="docutils literal notranslate"><span class="pre">lib.chipseq</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="lib.patterns_targets.html">Module <code class="docutils literal notranslate"><span class="pre">lib.patterns_targets</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="autodoc.html">Module documentation</a><ul>
      <li>Previous: <a href="lib.common.html" title="previous chapter">Module <code class="docutils literal notranslate"><span class="pre">lib.common</span></code></a></li>
      <li>Next: <a href="lib.patterns_targets.html" title="next chapter">Module <code class="docutils literal notranslate"><span class="pre">lib.patterns_targets</span></code></a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="module-lib-chipseq">
<h1>Module <code class="docutils literal notranslate"><span class="pre">lib.chipseq</span></code><a class="headerlink" href="#module-lib-chipseq" title="Permalink to this headline">¶</a></h1>
<p>Handling ChIP-seq peak-calling configuration correctly is complex. The
functions in this module help manipulate the config information so we can use
it more easily in the ChIP-seq workflow without cluttering the Snakefile.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#lib.chipseq.peak_calling_dict" title="lib.chipseq.peak_calling_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">peak_calling_dict</span></code></a>(config[,&nbsp;algorithm])</td>
<td>Returns a dictionary of peak-calling runs from the config.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#lib.chipseq.block_for_run" title="lib.chipseq.block_for_run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">block_for_run</span></code></a>(config,&nbsp;label,&nbsp;algorithm)</td>
<td>Returns the block for the (label, algorithm) run.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#lib.chipseq.samples_for_run" title="lib.chipseq.samples_for_run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">samples_for_run</span></code></a>(config,&nbsp;label,&nbsp;algorithm,&nbsp;…)</td>
<td>Returns the sample names configured for a particular peak-calling run</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#lib.chipseq.merged_input_for_ip" title="lib.chipseq.merged_input_for_ip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merged_input_for_ip</span></code></a>(sampletable,&nbsp;merged_ip)</td>
<td>Returns the merged input label for a merged IP label.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#lib.chipseq.detect_peak_format" title="lib.chipseq.detect_peak_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">detect_peak_format</span></code></a>(fn)</td>
<td>Figure out if a BED file is narrowPeak or broadPeak.</td>
</tr>
</tbody>
</table>
<div class="section" id="module-lib.chipseq">
<span id="details"></span><h2>Details<a class="headerlink" href="#module-lib.chipseq" title="Permalink to this headline">¶</a></h2>
<p>Helpers for ChIP-seq.</p>
<dl class="function">
<dt id="lib.chipseq.block_for_run">
<code class="descclassname">lib.chipseq.</code><code class="descname">block_for_run</code><span class="sig-paren">(</span><em>config</em>, <em>label</em>, <em>algorithm</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#block_for_run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#lib.chipseq.block_for_run" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the block for the (label, algorithm) run.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.6)"><em>dict</em></a>) – </li>
<li><strong>algorithm</strong> (<em>label</em><em>,</em>) – </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="lib.chipseq.detect_peak_format">
<code class="descclassname">lib.chipseq.</code><code class="descname">detect_peak_format</code><span class="sig-paren">(</span><em>fn</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#detect_peak_format"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#lib.chipseq.detect_peak_format" title="Permalink to this definition">¶</a></dt>
<dd><p>Figure out if a BED file is narrowPeak or broadPeak.</p>
<p>Returns None if undetermined.</p>
<p>This is useful for figuring out which autoSql file we should use or which
bigBed 6+4 or bigBed 6+3 format to use.</p>
</dd></dl>

<dl class="function">
<dt id="lib.chipseq.merged_input_for_ip">
<code class="descclassname">lib.chipseq.</code><code class="descname">merged_input_for_ip</code><span class="sig-paren">(</span><em>sampletable</em>, <em>merged_ip</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#merged_input_for_ip"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#lib.chipseq.merged_input_for_ip" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the merged input label for a merged IP label.</p>
<p>This is primarily used for the <cite>fingerprint</cite> rule, where we collect all the
available input BAMs together.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sampletable</strong> (<em>pandas.DataFrame</em>) – </li>
<li><strong>merged_ip</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.6)"><em>str</em></a>) – Label of IP to use, must be present in the <cite>label</cite> column of the
sampletable.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>This should make more sense if we have an example to work with…..</p>
<p>Samples ip1 and ip2 are technical replicates. They are from a different
experiment than ip3 and input3, hence their different
biological_material.</p>
<p>The way we know that input1 should be paired with ip1 and ip2 is because
it shares the same biological material.</p>
<p>Compare input1 and input9. They are not technical replicates (since they
do not share the same <cite>label</cite>) but they are biological replicates because
they share the same biological material.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">io</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="s1">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="s1">samplename  antibody   biological_material  label</span>
<span class="gp">... </span><span class="s1">ip1         gaf        s2cell-1             s2cell-gaf-1</span>
<span class="gp">... </span><span class="s1">ip2         gaf        s2cell-1             s2cell-gaf-1</span>
<span class="gp">... </span><span class="s1">ip3         ctcf       s2cell-2             s2cell-ctcf-1</span>
<span class="gp">... </span><span class="s1">input1      input      s2cell-1             s2cell-input-1</span>
<span class="gp">... </span><span class="s1">input3      input      s2cell-2             s2cell-input-3</span>
<span class="gp">... </span><span class="s1">input9      input      s2cell-1             s2cell-input-1&#39;&#39;&#39;</span><span class="p">),</span>
<span class="gp">... </span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;\s+&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merged_input_for_ip</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;s2cell-gaf-1&#39;</span><span class="p">)</span>
<span class="go">[&#39;s2cell-input-1&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merged_input_for_ip</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s1">&#39;s2cell-ctcf-1&#39;</span><span class="p">)</span>
<span class="go">[&#39;s2cell-input-3&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="lib.chipseq.peak_calling_dict">
<code class="descclassname">lib.chipseq.</code><code class="descname">peak_calling_dict</code><span class="sig-paren">(</span><em>config</em>, <em>algorithm=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#peak_calling_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#lib.chipseq.peak_calling_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of peak-calling runs from the config.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.6)"><em>dict</em></a>) – </li>
<li><strong>algorithm</strong> (<a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.6)"><em>None</em></a>) – If <cite>algorithm</cite> is None, dictionary is keyed by (label, algorithm).
Otherwise, only the runs for <cite>algorithm</cite> are returned, keyed by
<cite>label</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="lib.chipseq.samples_for_run">
<code class="descclassname">lib.chipseq.</code><code class="descname">samples_for_run</code><span class="sig-paren">(</span><em>config</em>, <em>label</em>, <em>algorithm</em>, <em>treatment</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/lib/chipseq.html#samples_for_run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#lib.chipseq.samples_for_run" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the sample names configured for a particular peak-calling run</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>config</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.6)"><em>dict</em></a>) – </li>
<li><strong>algorithm</strong> (<em>label</em><em>,</em>) – Used as keys into peak_calling_dict()</li>
<li><strong>treatment</strong> (<em>ip | input</em>) – </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.11</a>
      
      |
      <a href="_sources/lib.chipseq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>