
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>RNA-seq workflow &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Downstream RNA-seq in R" href="downstream-rnaseq.html" />
    <link rel="prev" title="References workflow" href="references.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Initial setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Running the tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="deploy.html">Deploying <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> and staying up-to-date</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">Downstream RNA-seq in R</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="external.html">“External” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="figures.html">“Figures” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocalization.html">Colocalization workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster.html">Running on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="autodoc.html">Module documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="references.html" title="previous chapter">References workflow</a></li>
      <li>Next: <a href="downstream-rnaseq.html" title="next chapter">Downstream RNA-seq in R</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="rna-seq-workflow">
<span id="rnaseq"></span><h1>RNA-seq workflow<a class="headerlink" href="#rna-seq-workflow" title="Permalink to this headline">¶</a></h1>
<p>This workflow is used for RNA-seq and RNA-seq-like analysis (like euRNA-seq,
RIP-seq or small RNA-seq).</p>
<p>This workflow can use references created by the references workflow with no
need to run the references workflow separately. This workflow performs the
following tasks:</p>
<blockquote>
<div><ul class="simple">
<li><p>Build a HISAT2 index</p></li>
<li><p>Build a salmon transcriptome index</p></li>
<li><p>Download a GTF annotation</p></li>
<li><p>Convert the GTF to refflat format</p></li>
<li><p>Trim reads with cutadapt</p></li>
<li><p>Align with HISAT2</p></li>
<li><p>Run FastQC on raw, trimmed, and aligned reads</p></li>
<li><p>Align reads to rRNA using bowtie2 to evaluate rRNA contamination</p></li>
<li><p>Count reads in genes with featureCounts</p></li>
<li><p>Run dupRadar and preseq to assess library complexity</p></li>
<li><p>Check for evidence of cross-contamination using fastq_screen on multiple
configured genomes</p></li>
<li><p>Assess transcript coverage with Picard CollectRnaSeqMetrics</p></li>
<li><p>Build bigWigs (optionally strand-specific) created from BAM files</p></li>
<li><p>Optionally merge bigWigs as defined by config</p></li>
<li><p>Aggregate QC results using MultiQC. Includes custom tables for library
sizes and rRNA contamination</p></li>
<li><p>Run various QC and differential expression. This is  performed in an
RMarkdown file that runs a standard DESeq2 differential expression
analysis along with diagnostic plots, exported tables of differentially
expressed genes for each contrast, and downstream GO analysis using
clusterProfiler. This file is run and rendered into an output HTML file.</p></li>
<li><p>Construct and upload a track hub of scaled coverage bigWigs for each
sample that can be viewed in UCSC Genome Browser</p></li>
</ul>
</div></blockquote>
<p>The DAG of jobs looks like this:</p>
<img alt="_images/rnaseq.png" src="_images/rnaseq.png" />
<p>To run the workflow on your own data:</p>
<ul class="simple">
<li><p>Install the dependencies (<a class="reference internal" href="getting-started.html#getting-started"><span class="std std-ref">Initial setup</span></a>)</p></li>
<li><p>Optionally run the workflow on provided test data (<a class="reference internal" href="tests.html#running-the-tests"><span class="std std-ref">Running the tests</span></a>)</p></li>
<li><p>Search the Snakefile for the string <code class="docutils literal notranslate"><span class="pre">NOTE</span></code> and edit rules as described by those comments</p></li>
<li><p>Edit the sampletable to reflect your sample IDs and their FASTQ files (<a class="reference internal" href="sampletable.html#sampletable"><span class="std std-ref">Sample tables</span></a>)</p></li>
<li><p>Edit the config file to reflect the references you want to use (<a class="reference internal" href="references-config.html#references-config"><span class="std std-ref">References config</span></a>)</p></li>
<li><p>Edit the config file to reflect the optional bigWig merging you would like (<a class="reference internal" href="config-yaml.html#cfg-merged-bigwigs"><span class="std std-ref">merged_bigwigs config section</span></a>)</p></li>
</ul>
<p>If you are running on a local machine, from the <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq</span></code> directory
and with the environment activated, this will do a dry run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake -n
</pre></div>
</div>
<p>If no errors are displayed, run the workflow (assuming 8 CPUs):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snakemake -j <span class="m">8</span> --use-conda
</pre></div>
</div>
<p>If you are running on a cluster other than NIH’s Biowulf, edit the settings in
the <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq/config/clusterconfig.yaml</span></code> to reflect the settings for
your cluster.</p>
<p>Edit <code class="docutils literal notranslate"><span class="pre">include/WRAPPER_SLURM</span></code> to reflect the environment name that you used.</p>
<p>From the <code class="docutils literal notranslate"><span class="pre">workflows/rnaseq</span></code> directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sbatch ../../include/WRAPPER_SLURM
</pre></div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/rnaseq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>