
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Running the tests &#8212; lcdb-wf 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Deploying lcdb-wf and staying up-to-date" href="deploy.html" />
    <link rel="prev" title="Initial setup" href="getting-started.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">lcdb-wf</a></h1>



<p class="blurb">Customizable workflows for high-throughput sequencing analysis</p>






<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting-started.html">Initial setup</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Running the tests</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#download-example-data">Download example data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#a-note-about-test-settings">A note about test settings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-rna-seq-workflow-with-example-data">Run the RNA-seq workflow with example data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-chip-seq-workflow-with-example-data">Run the ChIP-seq workflow with example data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-references-workflow-with-example-data">Run the references workflow with example data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#next-steps">Next steps</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="deploy.html">Deploying <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> and staying up-to-date</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Overview of workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="config.html">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnaseq.html">RNA-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream-rnaseq.html">Downstream RNA-seq in R</a></li>
<li class="toctree-l1"><a class="reference internal" href="chipseq.html">ChIP-seq workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="external.html">“External” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="figures.html">“Figures” workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="colocalization.html">Colocalization workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrappers.html">Wrappers</a></li>
<li class="toctree-l1"><a class="reference internal" href="cluster.html">Running on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a></li>
<li class="toctree-l1"><a class="reference internal" href="autodoc.html">Module documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">Developers</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="getting-started.html" title="previous chapter">Initial setup</a></li>
      <li>Next: <a href="deploy.html" title="next chapter">Deploying <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> and staying up-to-date</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="running-the-tests">
<span id="id1"></span><h1>Running the tests<a class="headerlink" href="#running-the-tests" title="Permalink to this headline">¶</a></h1>
<p>This section describes setting up for and running the example data, for when
you want to verify everything is working correctly. This reproduces the steps
that are performed during the automated tests on <a class="reference external" href="https:/circleci.com">Circle CI</a>. You can see the latest test results <a class="reference external" href="https://circleci.com/gh/lcdb/lcdb-wf/tree/master">here</a>.</p>
<p>The example run takes up about 360 MB of space and runs in about 15 mins on
2 cores.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The environment you created (see <a class="reference internal" href="getting-started.html#create-env"><span class="std std-ref">2. Create a new conda environment</span></a>) needs to be activated
for these steps.</p>
</div>
<div class="section" id="download-example-data">
<h2>Download example data<a class="headerlink" href="#download-example-data" title="Permalink to this headline">¶</a></h2>
<p>This will download the example data from our <a class="reference external" href="https://github.com/lcdb/lcdb-test-data">test data repository</a> into the directories
<code class="docutils literal notranslate"><span class="pre">workflows/{references,rnaseq,chipseq}/data</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python ci/get-data.py
</pre></div>
</div>
</div>
<div class="section" id="a-note-about-test-settings">
<span id="test-settings"></span><h2>A note about test settings<a class="headerlink" href="#a-note-about-test-settings" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The default configuration assumes a machine with large amounts of RAM.
Running the workflows as-is on a single machine with limited RAM may cause
all RAM to be consumed! Use <code class="docutils literal notranslate"><span class="pre">run_test.sh</span></code> as described below to avoid
this.</p>
</div>
<p>A major benefit of <code class="docutils literal notranslate"><span class="pre">lcdb-wf</span></code> is that the code undergoes automated testing on
<a class="reference external" href="https://circleci.com/gh/lcdb">CircleCI</a>. However this test environment only
has 2 cores and 2GB RAM. To accommodate this but to also allow the workflows to
run in their entirety in a reasonable time frame, we developed a small
representative <a class="reference external" href="https://github.com/lcdb/lcdb-test-data">test dataset</a> from
real-world data. We also need to make settings to the workflows, in particular
to set the Java VM memory to only 2GB for Java tools like Picard and FastQC.</p>
<p>We had to make a design decision: should the “default” state of the workflows
reflect production-ready (high-RAM) settings, or reflect test-ready (low RAM)
settings? We chose to have the default to be real-world, production-ready
settings, because we want to minimize the edits required (and therefore
possibility of introducing errors!) for running on real data.</p>
<p>What this all means is that if we want to run tests, we need to make some
adjustments. In each workflows directory, a <code class="docutils literal notranslate"><span class="pre">run_test.sh</span></code> script handles
this. This script runs a preprocessor, <code class="docutils literal notranslate"><span class="pre">ci/preprocessor.py</span></code>, which looks for
specially-formatted comments in the workflows, swaps out production settings
for test settings, and writes the results to a new <code class="docutils literal notranslate"><span class="pre">Snakefile.test</span></code> file that
is then run. In production, especially when running on a cluster, there’s no
need to do this.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">run_test.sh</span></code> simply passes all arguments on to Snakemake. Take a look at
the script to see what it’s doing, and see the examples below for usage.</p>
</div>
<div class="section" id="run-the-rna-seq-workflow-with-example-data">
<h2>Run the RNA-seq workflow with example data<a class="headerlink" href="#run-the-rna-seq-workflow-with-example-data" title="Permalink to this headline">¶</a></h2>
<p>With the <cite>lcdb-wf</cite> environment activated, change to the RNA-seq workflows
directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> workflows/rnaseq
</pre></div>
</div>
<p>First, run in dry-run mode which will print out the jobs to be run.  The
arguments will be described later, this is just to get things running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -n --use-conda
</pre></div>
</div>
<p>If all goes well, you will get lots of output ending with a summary of the
number of jobs that will be run. Then, use the same command but remove the
<code class="docutils literal notranslate"><span class="pre">-n</span></code>, and optionall include the <code class="docutils literal notranslate"><span class="pre">-j</span></code> argument to specify the number of
cores to use, for example <code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">8</span></code> if you have 8 cores on your machine (this
example just uses 2 cores):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -j <span class="m">2</span> --use-conda
</pre></div>
</div>
<p>This will take ~15 minutes to run.</p>
<p>Briefly,the file <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code> sets up lots of configuration like
where the fastq files are found, how to build references, and so on. The
workflow expects the file called <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code>; see <a class="reference internal" href="config.html#config"><span class="std std-ref">Configuration</span></a> for
more details. This workflow first imports the references workflow, which
downloads genome sequence and reference files and builds indexes as necessary
(HISAT2 genome index, salmon transcriptome index, bowtie2 index for rRNA, GTF
file of gene annotations) and then carries on with the RNA-seq workflow.</p>
<p>The RNA-seq workflow includes the standard mapping, counting, and differential
expression stages, as well as many quality-control steps. See <a class="reference internal" href="rnaseq.html#rnaseq"><span class="std std-ref">RNA-seq workflow</span></a> for
more details.</p>
<p>After the workflow runs, here are some useful points of interest in the output:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">data/rnaseq_samples/*</span></code>: sample-specific output. For example,
individual BAMs and bigWig files can be found here</li>
<li><code class="docutils literal notranslate"><span class="pre">data/aggregation/multiqc.html</span></code>:  MultiQC report.</li>
<li><code class="docutils literal notranslate"><span class="pre">downstream/rnaseq.html</span></code>: Differential expression results generated
from running the <code class="docutils literal notranslate"><span class="pre">downstream/rnaseq.Rmd</span></code> RMarkdown file.</li>
</ul>
</div></blockquote>
<p>See <a class="reference internal" href="rnaseq.html#rnaseq"><span class="std std-ref">RNA-seq workflow</span></a> for details.</p>
</div>
<div class="section" id="run-the-chip-seq-workflow-with-example-data">
<h2>Run the ChIP-seq workflow with example data<a class="headerlink" href="#run-the-chip-seq-workflow-with-example-data" title="Permalink to this headline">¶</a></h2>
<p>With the <cite>lcdb-wf</cite> environment activated, from the top-level directory of the
repo, change to the <code class="docutils literal notranslate"><span class="pre">workflows/chipseq</span></code> directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> workflows/chipseq
</pre></div>
</div>
<p>First, run in dry-run mode which will print out the jobs to be run.  The
arguments will be described later, this is just to get things running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -n --use-conda
</pre></div>
</div>
<p>If all goes well, you will get lots of output ending with a summary of the
number of jobs that will be run. Then, use the same command but remove the
<code class="docutils literal notranslate"><span class="pre">-n</span></code>, and optionall include the <code class="docutils literal notranslate"><span class="pre">-j</span></code> argument to specify the number of
cores to use, for example <code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">8</span></code> if you have 8 cores on your machine (this
example just uses 2 cores):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -j <span class="m">2</span> --use-conda
</pre></div>
</div>
<p>Like the RNA-seq workflow, the ChIP-seq workflow expects
a <code class="docutils literal notranslate"><span class="pre">config/config.yaml</span></code> file and includes the
<code class="docutils literal notranslate"><span class="pre">workflows/references/Snakemake</span></code> workflow, so that genome fastas are
downloaded and indexes built as necessary, before continuing on to the ChIP-seq
workflow. The ChIP-seq workflow includes QC, mapping, and peak-calling.</p>
<p>Points of interest:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">data/chipseq_samples/*</span></code>: sample-specific output. Individual BAM files
for a sample can be found here.</li>
<li><code class="docutils literal notranslate"><span class="pre">data/chipseq_merged/*</span></code>: technical replicates merged and re-deduped, or
if only one tech rep, symlinked to the BAM in the samples directory</li>
<li><code class="docutils literal notranslate"><span class="pre">data/chipseq_peaks/*</span></code>: peak-caller output, including BED files of
called peaks and bedGraph files of signal as output by each algorithm</li>
<li><code class="docutils literal notranslate"><span class="pre">data/chipseq_aggregation/multiqc.html</span></code>: MultiQC report</li>
</ul>
</div></blockquote>
<p>See <a class="reference internal" href="chipseq.html#chipseq"><span class="std std-ref">ChIP-seq workflow</span></a> for details.</p>
</div>
<div class="section" id="run-the-references-workflow-with-example-data">
<h2>Run the references workflow with example data<a class="headerlink" href="#run-the-references-workflow-with-example-data" title="Permalink to this headline">¶</a></h2>
<p>This is optional; parts of this workflow were actually run automatically as
needed for the RNA-seq and ChIP-seq workflows. However, running this workflow
on its own can be useful for setting up a new site, as it will build all
configured references the config file provided to it (as opposed to only
building the references specifically requested by either the ChIP-seq or
RNA-seq workflows).</p>
<p>From the top-level of the repo, change to the <code class="docutils literal notranslate"><span class="pre">workflows/references</span></code> directory:
.. code-block:: bash</p>
<blockquote>
<div>cd workflows/references</div></blockquote>
<p>First, run in dry-run mode which will print out the jobs to be run.  The
arguments will be described later, this is just to get things running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -n --use-conda --configfile ../../include/reference_configs/test.yaml
</pre></div>
</div>
<p>If all goes well, you will get lots of output ending with a summary of the
number of jobs that will be run. Then, use the same command but remove the
<code class="docutils literal notranslate"><span class="pre">-n</span></code>, and optionall include the <code class="docutils literal notranslate"><span class="pre">-j</span></code> argument to specify the number of
cores to use, for example <code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">8</span></code> if you have 8 cores on your machine (this
example just uses 2 cores):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./run_test.sh -j <span class="m">2</span> --use-conda --configfile ../../include/reference_configs/test.yaml
</pre></div>
</div>
<p>See <a class="reference internal" href="references.html#references"><span class="std std-ref">References workflow</span></a> for details.</p>
</div>
<div class="section" id="next-steps">
<h2>Next steps<a class="headerlink" href="#next-steps" title="Permalink to this headline">¶</a></h2>
<p>See <a class="reference internal" href="config.html#config"><span class="std std-ref">Configuration</span></a> for how to configure the workflows to work on your own data
and how to configure for your system.</p>
<p>See the <a class="reference internal" href="rnaseq.html#rnaseq"><span class="std std-ref">RNA-seq workflow</span></a>, <a class="reference internal" href="chipseq.html#chipseq"><span class="std std-ref">ChIP-seq workflow</span></a>, and <a class="reference internal" href="references.html#references"><span class="std std-ref">References workflow</span></a> sections for more
details on the above workflows, and then the <a class="reference internal" href="external.html#external"><span class="std std-ref">“External” workflow</span></a>, <a class="reference internal" href="figures.html#figures"><span class="std std-ref">“Figures” workflow</span></a>,
and <a class="reference internal" href="colocalization.html#colocalization"><span class="std std-ref">Colocalization workflow</span></a> sections for other workflows that can be used for
downstream analysis and integrating published data with newly-generated
results.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Ryan Dale, Justin Fear.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/tests.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>