---
title: Differential expression analysis
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r global_options, include=FALSE}
# Sets up global options for rendering RMarkdown into HTML.
knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE,
    cache.extra_file_dep_1 = file.info('../config/sampletable.tsv')$mtime,
    cache.extra_file_dep_2 = file.info('../data/rnaseq_aggregation/featurecounts.txt')$mtime,
    cache.extra_file_dep_2 = file.info('../data/rnaseq_samples/*/*.kallisto/abundance.h5')$mtime
)
```

```{r}
# Load the lcdbwf R package, which is stored locally.
# This package has many custom functions used throughout this document.
devtools::document('../../../lib/lcdbwf')
devtools::load_all('../../../lib/lcdbwf')
```

```{r config}

# HOW TO CONFIGURE ------------------------------------------------------
# Any chunks below that depend on config options should be cached and use one
# or more sections of the config object as arbitrary additional chunk options.
# By convention, we use the argument name "config" although there is nothing
# special about this name e.g.,
#
#      {r, cache=TRUE, config=config$annotation}
#
# Thereafter, any changes to config options under the "annotation" section will then
# invalidate that chunk's cache -- along with any other chunks that also have
# config$annotation included as a chunk option.
#
config <- lcdbwf::load_config('config.yaml')

# To keep this Rmd streamlined, much of the text is stored in a different file,
# text.yaml, and accessed via the `text` list after it is loaded here.
text <- yaml::yaml.load_file('text.yaml')

parallel <- config$parallel$parallel
if (config$parallel$parallel){
    register(MulticoreParam(config$parallel$cores))
}
```


# Changelog

**Initial results**

Last run: `r date()`


```{r}
library(AnnotationHub)
library(BiocParallel)
library(clusterProfiler)
library(cowplot)
library(DESeq2)
library(dplyr)
library(DT)
library(genefilter)
library(ggplot2)
library(gridExtra)
library(plotly)
library(purrr)
library(ReactomePA)
library(readr)
library(reshape)
library(tibble)
library(tximport)
library(UpSetR)
```



```{r coldata_setup}
# Set up all of the metadata for the samples and experimental design. Use this
# chunk to modify if needed.
colData <- read.table(config$main$sampletable, sep='\t', header=TRUE, stringsAsFactors=FALSE)

# lcdb-wf requires that the first column of the sampletable contains sample
# names. Use that to set the rownames.
rownames(colData) <- colData[,1]
```

```{r dds_initial, cache=TRUE, config=c(config$main, config$toggle$salmon, config$toggle$kalliso)}
# Convert featureCounts gene-level counts into DESeq2 object, and run
# variance-stabiliizing transform.
dds_initial <- lcdbwf::make_dds(
  list(sampletable=colData, design=~1),
  config=config,
  parallel=config$parallel$parallel
)
vsd <- varianceStabilizingTransformation(dds_initial, blind=TRUE)
```

# Experiment overview

Here is the sample table with metadata used for this analysis:

```{r}
exclude.for.printing <- c('featurecounts.path', 'salmon.path', 'kallisto.path',
                          'orig_filename', 'orig_filename_R2', 'layout',
                          'sizeFactor')
colData(dds_initial) %>%
    as.data.frame() %>%
    dplyr::select(-contains(exclude.for.printing)) %>%
    datatable()
```


# Sample similarity and QC

## Clustered heatmap

The following heatmap shows a hierarchical clustering of pairwise distances
between samples. Darker blue means less distant (i.e. more similar). In general
we expect to see replicates clustering together and separation of treatments.

```{r sample_heatmap, cache=TRUE, config=config$plotting$covariates_for_plots, dependson='dds_initial'}
# Plot a clustered heatmap of sample distances, for QC.
lcdbwf::plot_heatmap(vsd, colData, cols_for_grouping=config$plotting$covariates_for_plots)
```

## PCA {.tabset}

Another way of looking at sample clustering is principal components analysis
(PCA). The x- and y-axes do not have units, rather, they represent the
dimensions along which the samples vary the most. The amount of variance
explained by each principal component is indicated in the axes label.


```{r pca, results='asis', cache=TRUE, config=config$plotting$covariates_for_plots, dependson='dds_initial'}
# Plot PCA plots, optionally tabbed with different factors colored, for QC.
for(group in config$plotting$covariates_for_plots){
    lcdbwf::mdcat('### ', paste(group, sep=", "))
    p <- lcdbwf::plotPCA.ly(vsd, intgroup=group)
    print(htmltools::tagList(ggplotly(p)))
}
```

```{r, results='asis', eval=!(config$toggle$salmon | config$toggle$kallisto)}
# Note that when loading Salmon or Kallisto, DESeq2 does not calculate size
# factors.

mdcat(text$sizefactors)
mdcat("### Size factors")
p <- lcdbwf::sizefactors_barplot(dds_initial)
ggplotly(p)

mdcat("### Size factors vs total read count")
p <- lcdbwf::sizefactors_vs_total(dds_initial)
ggplotly(p)
```

```{r dds_list, cache=TRUE, config=config$main}
# Create a list of dds objects used as the basis for differential expression.
#
# May need substantial editing.
lst <- list(
            # NOTE: Delete these for real data! They are just examples to
            # demonstrate functionality.
            # ------------------------------------------------------------------
            main=list(
                sampletable=colData,
                design=~group),
            no_rep4=list(
                sampletable=colData %>% filter(samplename != 'sample4'),
                design=~group,
                subset.counts=TRUE
            ),
            salmon=list(
                sampletable=colData,
                design=~group,
                salmon=TRUE),
            kallisto=list(
                sampletable=colData,
                design=~group,
                kallisto=TRUE)
            # ------------------------------------------------------------------
          )
dds_list <- map(lst, lcdbwf::make_dds, config=config, parallel=config$parallel$parallel)
```


```{r, results='asis', cache=TRUE, dependson='dds_list', config=config$toggle$dds_diagnostics}
if (config$toggle$dds_diagnostics){ lcdbwf::dds_diagnostics(dds_list, text) }
```

# Differential expression {.tabset}

Now we perform differential expression analysis for the contrasts listed in the
table below.

Under the table are tabs, one for each contrast. Under each contrast's tab,
there are more tabs for the various output.


```{r, results='asis'}
lcdbwf::folded_markdown(text$results_table, "Help for table")
```


```{r results_01, dependson='dds_list', cache=TRUE}
# Perform differential expression testing.
#
# To take advantage of caching:
#   - create each list in a different chunk
#   - ensure chunk name starts with "results_"
#   - ensure list name starts with "contr_". The rest of the name will be used
#     as a label to sort results and create output files.
#
# This will need substantial editing.
contr_main <- list(
    res=lfcShrink(
            dds_list[['main']],
            contrast=c('group', 'treatment', 'control'),
            type='normal',
            parallel=parallel
        ),
        dds='main',
        label='Using a log2FoldChange threshold of 0'
)
```

```{r results_02, dependson='dds_list', cache=TRUE}
# NOTE: Example contrast #2----------------------------------------------------
#    Using the example data, this compares treatment group to control group but
#    requiring genes to have >4-fold differences (log2(4) = 2). Change to
#    reflect your experiment.
#    If using an lfcThreshold, must update elsewhere throughout the code.
res.lfcthresh.2 <- results(
    dds_list[['main']],
    contrast=c('group', 'treatment', 'control'),
    lfcThreshold=2)

contr_lfc2 <- list(
    res=lfcShrink(
        dds_list[['main']],
        contrast=c('group', 'treatment', 'control'),
        parallel=parallel,
        res=res.lfcthresh.2,
        type='normal'
    ),
    dds='main',
    label='Using a log2FoldChange threshold of >2'
)
```

```{r results_03, dependson='dds_list', cache=TRUE}
contr_salmon <- list(
  res=lfcShrink(
    dds_list[['salmon']],
    contrast=c('group', 'treatment', 'control'),
    type='normal',
    parallel=parallel),
  dds='salmon',
  label='Using Salmon')
```

```{r results_04, dependson='dds_list', cache=TRUE}
contr_kallisto <- list(
  res=lfcShrink(
    dds_list[['kallisto']],
    contrast=c('group', 'treatment', 'control'),
    type='normal',
    parallel=parallel),
  dds='kallisto',
  label='Using Kallisto')
```


```{r assemble_variables, cache=TRUE, config=config$annotation, dependson=knitr::all_labels()[grepl("^results_", knitr::all_labels())]}
res_list <- lcdbwf::collect_objects("^contr_")
res_list <- lcdbwf::attach_extra(res_list, config)
```

```{r, results='asis'}
summarize_res_list(res_list) %>% datatable
```

```{r reportresults, results='asis', config=config$annotation, dependson='assemble_variables', cache=TRUE}
# Create nested tabs showing the results.
# First level of tabs will be labels of results; second level will be different
# output (MA plot, volcano plot, etc)

for (name in names(res_list)){
  dds_i <- dds_list[[res_list[[name]][['dds']] ]]
  res_i <- res_list[[name]][['res']]
  label <- res_list[[name]][['label']]
  genes_to_label <- lcdbwf::genes_to_label(res_i, n=5, config)
  lcdbwf::mdcat('## ', label, ' {.tabset}')

  # TODO:
  # https://hdgitappip01.nichd.nih.gov/bspc/core/labeled-ma-plot/-/blob/master/plotMA.label.R
  # is the latest version

  lcdbwf::mdcat('### M-A plot')
  lcdbwf::folded_markdown(text$results_plots$ma, "Help")
  print(lcdbwf::plotMA_label(
    res_i,
    genes_to_label=genes_to_label,
    label_column=config$annotation$label_column))

  lcdbwf::mdcat('### Volcano plot')
  lcdbwf::folded_markdown(text$results_plots$volcano, "Help")
  print(lcdbwf::plot_volcano_label(
    res_i,
    genes_to_label=genes_to_label,
    label_column=config$annotation$label_column))

  lcdbwf::mdcat('### P-value distribution')
  lcdbwf::folded_markdown(text$results_plots$pval_hist, "Help")
  lcdbwf::pval_hist(res_i)

  if (config$toggle$results_diagnostics){
    lcdbwf::results_diagnostics(res=res_i, dds=res_list[[name]]$dds, name=name, config=config, text=text)
  }

}
```

```{r upsetplots, results='asis', eval=length(res_list)>1, config=config$annotation, dependson='selections'}
lcdbwf::mdcat(text$upset_plots)

# Compare overlap across multiple differential expression contrasts.
lcdbwf::plot_upsets(res_list, label_column=config$annotation$label_column)
```

# Exported results

Single Excel file with worksheets for each contrast: [results/consolidated_results.xlsx](results/consolidated_results.xlsx)

Alternatively, the files below are TSVs that can be opened in Excel or used
progammatically with downstream tools:

```{r, results='asis'}
# Write out files for full and each selection, and create a link to them in the
# HTML generated by this RMarkdown.
tbl <- lcdbwf::exported_tsvs(res_list)
knitr::kable(tbl, row.names=FALSE)
```

```{r, results='asis'}
lcdbwf::exported_excel(res_list, dds_list, file='results/consolidated_results.xlsx')
```


```{r combined_rds, dependson='assemble_variables'}
obj <- compose_results(res_list, dds_list)
saveRDS(obj, file='combined.Rds')
```


# Session info
For reproducibility purposes, here is the output of `sessionInfo()` showing the
versions of all packages used here.

```{r}
sessionInfo()
```
